### PR TITLE
feat: publish artifact to reuse the build output

### PR DESCRIPTION
**Pull request by fpaul-1A**

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -32,6 +32,13 @@ jobs:
         run: yarn build:swagger-gen
       - name: Build
         run: yarn build
+      - name: zip dist content
+        run: zip -q -r dist.zip . -i "*/dist/*" -x "*/node_modules/*" ".cache/*"
```
Comment (kpanot) [inline .github/workflows/main.yml:36]: Could be more specific with the regexp form `workspaces` field: `apps/*/dist`, `packages/*/*/dist` and `packages/@o3r/amaterasu/*/dist`

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -51,13 +51,14 @@ jobs:
...
         uses: ./.github/actions/setup
+      - name: Download dist folders
+        uses: actions/download-artifact@v3
+        if: github.event.release == null
         with:
-          enable-build-cache: 'true'
-          install-jdk: 'true'
-      - name: Build Swagger CodeGen
-        run: yarn build:swagger-gen
-      - name: Build
```
Comment (kpanot) [inline .github/workflows/publish.yml:59]: you still need to build in case of `release` trigger

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -51,13 +51,14 @@ jobs:
...
-        run: yarn build
+          name: dist
+          path: '.'
+      - name: Unzip dist
```
Comment (kpanot) [inline .github/workflows/publish.yml:60]: Should not be run when `release` trigger

**Comment by kpanot** [comment]
I think you should move out the `release` trigger to another pipeline dedicated to it.
Both event triggers seems to diverge quite a significantly

**Comment by fpaul-1A** [comment]
> I think you should move out the `release` trigger to another pipeline dedicated to it. Both event triggers seems to diverge quite a significantly

done

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -17,21 +17,15 @@ on:
...
-        uses: ./.github/actions/setup
+      - uses: ./.github/actions/setup
         with:
           enable-build-cache: 'true'
           install-jdk: 'true'
-      - name: Build Swagger CodeGen
-        run: yarn build:swagger-gen
-      - name: Build
-        run: yarn build
+      - run: yarn build:swagger-gen
```
Comment (kpanot) [inline .github/workflows/main.yml:26]: why did you remove the task names?

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -49,17 +47,9 @@ jobs:
...
-          install-jdk: 'true'
-      - name: Build Swagger CodeGen
-        run: yarn build:swagger-gen
-      - name: Build
-        run: yarn build
-      - name: Set new version
-        run: yarn set:version ${{ steps.get-version.outputs.version }}
+      - uses: ./.github/actions/setup
+      - uses: ./.github/actions/download-build-output
+      - run: yarn set:version ${{ steps.get-version.outputs.version }}
```
Comment (kpanot) [inline .github/workflows/publish.yml:52]: you can change the `is-prerelease` (in following lines) to `is-prerelease: ${{ inputs.prerelease }}`

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,72 @@
+name: Release Packages
+
+on:
+  workflow_call:
```
Comment (kpanot) [inline .github/workflows/release.yml:4]: this trigger is not needed on this workflow

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+          enable-build-cache: 'false'
+          install-jdk: 'true'
+      - run: yarn build:swagger-gen
+      - run: yarn build
+      - run: yarn set:version ${{ inputs.version }}
+      - name: Get tag name
+        id: get-npm-tag
+        uses: ./.github/actions/get-npm-tag
+        with:
+          is-prerelease: ${{ inputs.prerelease || (github.event.release && github.event.release.prerelease) }}
```
Comment (kpanot) [inline .github/workflows/release.yml:54]: or `is-prerelease: ${{ github.event.release && github.event.release.prerelease }}`

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      - name: Publish
+        run: yarn run publish --tag=${{ steps.get-npm-tag.outputs.tag }} --always-auth=true
+        env:
+          GITHUB_AUTH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+          AZURE_AUTH_TOKEN: ${{ secrets.AZURE_AUTH_TOKEN }}
+          AZURE_USERNAME: ${{ secrets.AZURE_USERNAME }}
+          AZURE_EMAIL: ${{ secrets.AZURE_EMAIL }}
+      - name: Publish Extensions
+        if: '!inputs.prerelease'
```
Comment (kpanot) [inline .github/workflows/release.yml:65]: `if: github.event.release && !github.event.release.prerelease`

File: .github/workflows/main.yml (hunk 4 above)
Comment (fpaul-1A) [reply .github/workflows/main.yml:26]: I removed it because I think that `Build` is not necessarily a better name than the default `Run yarn build`

**Review by jbourgeois-1A — APPROVED** [review]
That's a lot less than 4mn when we upload only one file instead of 4000 😄

File: .github/workflows/release.yml (hunk 8 above)
Comment (fpaul-1A) [reply .github/workflows/release.yml]: done

File: .github/workflows/release.yml (hunk 7 above)
Comment (fpaul-1A) [reply .github/workflows/release.yml]: done

File: .github/workflows/release.yml (hunk 6 above)
Comment (fpaul-1A) [reply .github/workflows/release.yml]: removed

File: .github/workflows/publish.yml (hunk 5 above)
Comment (fpaul-1A) [reply .github/workflows/publish.yml]: done

**Review by kpanot — DISMISSED** [review]

File: .github/actions/download-build-output/action.yml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+name: Download Build Output
+description: Download and restore the `dist` folders from github artifact (check ../upload-build-output)
+
+runs:
+  using: 'composite'
+  steps:
+    - uses: actions/download-artifact@v3
+      if: github.event.release == null
```
Comment (kpanot) [inline .github/actions/download-build-output/action.yml:8]: should not be part of the action (which is not used in case of `release` trigger)

Comment (fpaul-1A) [reply .github/actions/download-build-output/action.yml]: removed

**Review by jbourgeois-1A — APPROVED** [review]

**Review by cpaulve-1A — APPROVED** [review]
